### PR TITLE
Slowly port VAR rebased changes in

### PR DESCRIPTION
**Pull request by bashtage**
Bring in the changes in `vector_ar` and dependent modules only.  Manually ported in fixes from #3028 rather than trying to rebase the entire PR.  

Lots of doc string changes.
A few external variables rename all with deprecation.
Structured commits that should each pass the test suite.

xref #5206  #3028

Plan is to get #3028 into VAR over a number of PRs to limit reviews work (which is already a lot).

### CODE REVIEW

File: statsmodels/tsa/vector_ar/irf.py [hunk 1]
```diff
@@ -57,7 +57,7 @@ def __init__(self, model, P=None, periods=10, order=None, svar=False,
         if svar:
             self.svar_irfs = model.svar_ma_rep(periods, P=P)
         else:
-            self.orth_irfs = model.orth_ma_rep(periods, P=P)
+            self.orth_irfs = model.orth_ma_rep(periods, p=P)
```
Comment (jbrockmendel) [inline statsmodels/tsa/vector_ar/irf.py]: should this also be done in the `svar_ma_rep` above?

File: statsmodels/tsa/vector_ar/irf.py [hunk 2]
```diff
@@ -296,11 +296,11 @@ def errband_mc(self, orth=False, svar=False, repl=1000,
         model = self.model
         periods = self.periods
         if svar == True:
```
Comment (jbrockmendel) [inline statsmodels/tsa/vector_ar/irf.py]: `if svar`

File: statsmodels/tsa/vector_ar/irf.py [hunk 3]
```diff
@@ -394,8 +394,8 @@ def err_band_sz2(self, orth=False, svar=False, repl=1000, signif=0.05,
         periods = self.periods
         irfs = self._choose_irfs(orth, svar)
         neqs = self.neqs
-        irf_resim = model.irf_resim(orth=orth, repl=repl, T=periods, seed=seed,
-                                   burn=100)
+        irf_resim = model.irf_resim(orth=orth, repl=repl, horizon=periods, seed=seed,
+                                    burn=100)
```
Comment (jbrockmendel) [inline statsmodels/tsa/vector_ar/irf.py]: Should the `burn` kwarg in the method be passed instead of hard-coding 100?

Comment (jbrockmendel) [reply statsmodels/tsa/vector_ar/irf.py]: ditto below

File: statsmodels/tsa/vector_ar/irf.py [hunk 4]
```diff
@@ -457,8 +457,8 @@ def err_band_sz3(self, orth=False, svar=False, repl=1000, signif=0.05,
         periods = self.periods
         irfs = self._choose_irfs(orth, svar)
         neqs = self.neqs
-        irf_resim = model.irf_resim(orth=orth, repl=repl, T=periods, seed=seed,
-                                   burn=100)
+        irf_resim = model.irf_resim(orth=orth, repl=repl, horizon=periods, seed=seed,
```
Comment (jbrockmendel) [inline statsmodels/tsa/vector_ar/irf.py]: I think all uses of irf_resim are here.  Should that just be an irf method instead of a VARResults method?

File: statsmodels/tsa/vector_ar/irf.py [hunk 5]
```diff
@@ -639,8 +639,9 @@ def cum_errband_mc(self, orth=False, repl=1000,
         """
         model = self.model
         periods = self.periods
-        return model.irf_errband_mc(orth=orth, repl=repl,
-                                    T=periods, signif=signif, seed=seed, burn=burn, cum=True)
+        return model.irf_errband_mc(orth=orth, repl=repl, horizon=periods,
+                                    signif=signif, seed=seed, burn=burn,
+                                    cum=True)
```
Comment (jbrockmendel) [inline statsmodels/tsa/vector_ar/irf.py]: I wouldn't be opposed to changing `cum` to `cumul`

File: statsmodels/tsa/vector_ar/svar_model.py [hunk 6]
```diff
@@ -12,6 +12,10 @@
 import numpy as np
 import numpy.linalg as npl
 from numpy.linalg import slogdet
+try:
+    from pandas.util._decorators import deprecate_kwarg
+except ImportError:
+    from pandas.util.decorators import deprecate_kwarg
```
Comment (jbrockmendel) [inline statsmodels/tsa/vector_ar/svar_model.py]: +1 for using this

File: statsmodels/tsa/vector_ar/svar_model.py [hunk 7]
```diff
@@ -12,6 +12,10 @@
 import numpy as np
 import numpy.linalg as npl
```
Comment (jbrockmendel) [inline statsmodels/tsa/vector_ar/svar_model.py]: Getting rid of this and other unique `import as [...]` names in this modules would be nice

File: statsmodels/tsa/vector_ar/svar_model.py [hunk 8]
```diff
@@ -465,18 +469,19 @@ class SVARProcess(VARProcess):
...
         self.coefs = coefs
         self.intercept = intercept
-        self.sigma_u = sigma_u
+        self.cov_resid = cov_resid
         self.A_solve = A_solve
         self.B_solve = B_solve
         self.names = names
 
-    def orth_ma_rep(self, maxn=10, P=None):
+    def orth_ma_rep(self, maxn=10, p=None):
```
Comment (jbrockmendel) [inline statsmodels/tsa/vector_ar/svar_model.py]: `svar_ma_rep` below also change `P`-->`p`?

Actually, why this change?  `P` is a matrix right?

File: statsmodels/tsa/vector_ar/tests/test_var.py [hunk 9]
```diff
@@ -378,15 +382,15 @@ def test_acorr(self):
         acorrs = self.res.acorr(10)
 
     def test_forecast(self):
-        point = self.res.forecast(self.res.y[-5:], 5)
+        point = self.res.forecast(self.res.endog[-5:], 5)
```
Comment (jbrockmendel) [inline statsmodels/tsa/vector_ar/tests/test_var.py]: Very much +1 for getting rid of the multiple aliases with the `# keep alias for now`s

File: statsmodels/tsa/vector_ar/tests/test_var.py [hunk 10]
```diff
@@ -378,15 +382,15 @@ def test_acorr(self):
...
 
     def test_forecast_interval(self):
-        y = self.res.y[:-self.p:]
+        y = self.res.endog[:-self.p:]
```
Comment (jbrockmendel) [inline statsmodels/tsa/vector_ar/tests/test_var.py]: Maybe I need more caffeine; what does this slice mean?

File: statsmodels/tsa/vector_ar/tests/test_var.py (hunk 9 above)
Comment (jbrockmendel) [reply statsmodels/tsa/vector_ar/tests/test_var.py]: https://github.com/jbrockmendel/sm2/blob/master/sm2/tools/decorators.py#L12

File: statsmodels/tsa/vector_ar/var_model.py [hunk 11]
```diff
@@ -4,49 +4,61 @@
...
+                                       iteritems)
+
 from collections import defaultdict
 
 import numpy as np
-import scipy.stats as stats
+try:
+    from pandas.util._decorators import deprecate_kwarg
+except ImportError:
+    from pandas.util.decorators import deprecate_kwarg
```
Comment (jbrockmendel) [inline statsmodels/tsa/vector_ar/var_model.py]: Should this be done in compat.pandas?

File: statsmodels/tsa/vector_ar/var_model.py [hunk 12]
```diff
@@ -57,165 +69,155 @@ def ma_rep(coefs, maxn=10):
...
                 break
 
-            phis[i] += np.dot(phis[i-j], coefs[j-1])
+            phis[i] += np.dot(phis[i - j], coefs[j - 1])
 
     return phis
 
 
-def is_stable(coefs, verbose=False):
+@deprecate_kwarg('verbose', 'eigenvalues')
```
Comment (jbrockmendel) [inline statsmodels/tsa/vector_ar/var_model.py]: Why this change?  `verbose` seems clearer to me.

File: statsmodels/tsa/vector_ar/tests/test_var.py (hunk 10 above)
Comment (bashtage) [reply statsmodels/tsa/vector_ar/tests/test_var.py]: The last colon is unnecessary

File: statsmodels/tsa/vector_ar/svar_model.py (hunk 8 above)
Comment (bashtage) [reply statsmodels/tsa/vector_ar/svar_model.py]: `P` is an ideal python variable name, `p` is.

File: statsmodels/tsa/vector_ar/var_model.py [hunk 13]
```diff
@@ -57,165 +69,155 @@ def ma_rep(coefs, maxn=10):
...
+def is_stable(coefs, eigenvalues=False):
```
Comment (jbrockmendel) [inline statsmodels/tsa/vector_ar/var_model.py]: Several of the functions in this module are redundant with methods in the classes.  `is_stable` is one that ideally I'd like to see in a `VARParams`-like class, which would be inherited by `VARProcess`.  `VARParams` would basically just have `coefs` attribute.

**Comment by bashtage** [comment]
I'm trying to have 1 file at a time plus any required fixes in other files.

These should all get done before the next release I would hope, but don't want to push a huge PR (this is already big) that has ~0% chance of getting in.

File: statsmodels/tsa/vector_ar/var_model.py (hunk 12 above)
Comment (jbrockmendel) [reply statsmodels/tsa/vector_ar/var_model.py]: Oh I see; verbose and eigenvalues are doing different things.  In that case is deprecate_kwarg the right pattern?

File: statsmodels/tsa/vector_ar/var_model.py [hunk 14]
```diff
@@ -57,165 +69,155 @@ def ma_rep(coefs, maxn=10):
...
+    Notes
+    -----
+    Computes the eigenvalues of the coefficient matrix in the the companion
+    form of the VAR(p).
+    """
+    a_var1 = util.comp_matrix(coefs)
+    eigs = np.linalg.eigvals(a_var1)
+    stable = bool((np.abs(eigs) <= 1).all())
+    if eigenvalues:
+        return stable, eigs
```
Comment (jbrockmendel) [inline statsmodels/tsa/vector_ar/var_model.py]: -1 for multiple return

File: statsmodels/tsa/vector_ar/var_model.py [hunk 15]
```diff
@@ -485,21 +604,20 @@ def summary(self):  # basically copied from (now deleted) print_ic_table()
...
     def __str__(self):
         return "<" + self.__module__ + "." + self.__class__.__name__ \
-                   + " object. Selected orders are: AIC -> " + str(self.aic) \
-                   + ", BIC -> " + str(self.bic)  \
-                   + ", FPE -> " + str(self.fpe) \
-                   + ", HQIC -> " + str(self.hqic) + ">"
+               + " object. Selected orders are: AIC -> " + str(self.aic) \
+               + ", BIC -> " + str(self.bic) \
+               + ", FPE -> " + str(self.fpe) \
+               + ", HQIC -> " + str(self.hqic) + ">"
```
Comment (jbrockmendel) [inline statsmodels/tsa/vector_ar/var_model.py]: Parens instead of "\"?

File: statsmodels/tsa/vector_ar/var_model.py (hunk 13 above)
Comment (bashtage) [reply statsmodels/tsa/vector_ar/var_model.py]: The class heirarchy in VAR/VARProcess doesn't make sense.  VARPRocess needs to know about exog stuff which isn't actually in VARProcess.  It could use a big refactor but this is more than I want to try in this PR.  Ditto some of the functions (not in classes) that have `self` as the first argument so they can be used somewhere else. These should be rewritten as propper functions with lists of arguments rather than relying on a particular structure to self.

File: statsmodels/tsa/vector_ar/var_model.py (hunk 15 above)
Comment (bashtage) [reply statsmodels/tsa/vector_ar/var_model.py]: Needs a rewrite.

File: statsmodels/tsa/vector_ar/var_model.py (hunk 14 above)
Comment (jbrockmendel) [reply statsmodels/tsa/vector_ar/var_model.py]: Like many of the pep8 improvements, this is not a big deal in itself, but as a pattern around the codebase adds up.

File: statsmodels/tsa/vector_ar/var_model.py [hunk 16]
```diff
@@ -808,35 +984,44 @@ def __init__(self, coefs, coefs_exog, sigma_u, names=None, _params_info=None):
             self.intercept = np.zeros(self.neqs)
 
     def get_eq_index(self, name):
```
Comment (jbrockmendel) [inline statsmodels/tsa/vector_ar/var_model.py]: This is never actually used.  There's a test in test_var and a commented-out usage in irf.  Remove/privatize?

File: statsmodels/tsa/vector_ar/var_model.py [hunk 17]
```diff
@@ -1297,23 +1670,66 @@ def resid_acorr(self, nlags=1):
...
+
+        Notes
+        -----
+        Deprecated. Use cov_resid
+        """
+        # Deprecated in 0.10.0, remove in 0.11.0
+        import warnings
+        warnings.warn('sigma_u_mle has been deprecated in favor of '
+                      'cov_resid_mle', DeprecationWarning)
+        return self.cov_resid
```
Comment (jbrockmendel) [inline statsmodels/tsa/vector_ar/var_model.py]: Should return `self.cov_resid_mle`?  Is the docstring right?

File: statsmodels/tsa/vector_ar/var_model.py [hunk 18]
```diff
@@ -1322,67 +1738,78 @@ def cov_params(self):
...
+        return np.kron(scipy.linalg.inv(np.dot(z.T, z)), self.cov_resid)
 
     def cov_ybar(self):
+        """
+        Deprecated.  Use cov_sample_mean.
+        """
+        return self.cov_sample_mean()
+
+    def cov_sample_mean(self):
         r"""Asymptotically consistent estimate of covariance of the sample mean
```
Comment (jbrockmendel) [inline statsmodels/tsa/vector_ar/var_model.py]: "covariance of the sample mean" seems misleading.  The sample mean is known.  Maybe the covariance of the estimated process mean?

File: statsmodels/tsa/vector_ar/var_model.py [hunk 19]
```diff
@@ -1558,21 +2083,25 @@ def irf_resim(self, orth=False, repl=1000, T=10,
...
+            ret = VAR(simulation, exog=self.exog).fit(maxlags=k_ar,
+                                                      trend=self.trend)
+            if orth:
+                rep = ret.orth_ma_rep(maxn=horizon)
+            else:
+                rep = ret.ma_rep(maxn=horizon)
+            return rep.cumsum(axis=0) if cum else rep
 
         for i in range(repl):
             # discard first hundred to eliminate correct for starting bias
```
Comment (jbrockmendel) [inline statsmodels/tsa/vector_ar/var_model.py]: hundred --> burn

File: statsmodels/tsa/vector_ar/var_model.py [hunk 20]
```diff
@@ -1985,139 +2596,220 @@ def plot_acorr(self, nlags=10, resid=True, linewidth=8):
...
 
-        # See Lutkepohl pp. 146-150
+        # See Lütkepohl pp. 146-150
 
         aic = ld + (2. / nobs) * free_params
         bic = ld + (np.log(nobs) / nobs) * free_params
         hqic = ld + (2. * np.log(np.log(nobs)) / nobs) * free_params
-        fpe = ((nobs + self.df_model) / self.df_resid) ** neqs * np.exp(ld)
+        fpe = ((nobs + self.df_model) / self.df_resid)**neqs * np.exp(ld)
+        res = Bunch()
```
Comment (jbrockmendel) [inline statsmodels/tsa/vector_ar/var_model.py]: -1

File: statsmodels/tsa/vector_ar/var_model.py [hunk 21]
```diff
@@ -2133,9 +2825,9 @@ def summary(self):
...
         rng = lrange(self.periods)
         for i in range(self.neqs):
-            ppm = output.pprint_matrix(self.decomp[i], rng, self.names)
+            ppm = output.pprint_matrix(self.decomp[i], rng, self.endog_names)
 
-            buf.write('FEVD for %s\n' % self.names[i])
+            buf.write('FEVD for %s\n' % self.endog_names[i])
             buf.write(ppm + '\n')
 
         print(buf.getvalue())
```
Comment (jbrockmendel) [inline statsmodels/tsa/vector_ar/var_model.py]: print needed?

File: statsmodels/compat/pandas.py [hunk 22]
```diff
@@ -33,3 +33,9 @@
 assert_frame_equal = testing.assert_frame_equal
 assert_index_equal = testing.assert_index_equal
 assert_series_equal = testing.assert_series_equal
+
+try:
+    from pandas.util._decorators import deprecate_kwarg
+except ImportError:
+    # 0.18 and earlier
```
Comment (jbrockmendel) [inline statsmodels/compat/pandas.py]: min version is already 0.19

**Comment by jbrockmendel** [comment]
Big PR, generally very positive.  Will make another pass before long.

**Comment by jbrockmendel** [comment]
VAR and SVAR have a `_model_type` attribute that should go

File: statsmodels/tsa/vector_ar/var_model.py (hunk 13 above)
Comment (josef-pkt) [reply statsmodels/tsa/vector_ar/var_model.py]: I'm not sure exog belong in VARProcess, or at least being able to keep it out as option.
VARProcess should be the dynamic properties of the underlying VAR process, and not so much about exogenous changes in mean.

File: statsmodels/tsa/vector_ar/var_model.py [hunk 23]
```diff
@@ -1643,48 +2192,124 @@ def irf(self, periods=10, var_decomp=None, var_order=None):
...
             order = order_new
         return _reordered(self, order)
 
-    # --------------------------------------------------------------------------
-    # VAR Diagnostics: Granger-causality, whiteness of residuals, normality, etc
+    # ------------------------------------------------------------------------
+    # VAR Diagnostics: Granger-causality, whiteness of resids, normality, etc.
+
+    # TODO: Rescue?
+    # def test_causality_all(self, kind='F', signif=0.05):
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply statsmodels/tsa/vector_ar/var_model.py]: This is from #3028

File: statsmodels/compat/pandas.py (hunk 22 above)
Comment (bashtage) [reply statsmodels/compat/pandas.py]: 0.19 it turns out

File: statsmodels/tsa/vector_ar/var_model.py (hunk 18 above)
Comment (bashtage) [reply statsmodels/tsa/vector_ar/var_model.py]: It is the covariance of the sample mean estimator - cov_sample_mean_estimator seems excessive,

File: statsmodels/tsa/vector_ar/var_model.py (hunk 17 above)
Comment (bashtage) [reply statsmodels/tsa/vector_ar/var_model.py]: Yes.

File: statsmodels/tsa/vector_ar/var_model.py (hunk 16 above)
Comment (bashtage) [reply statsmodels/tsa/vector_ar/var_model.py]: Trying to minimize changes to api now unless I fee strongly about it.

File: statsmodels/tsa/vector_ar/var_model.py (hunk 20 above)
Comment (bashtage) [reply statsmodels/tsa/vector_ar/var_model.py]: I don't like the name 'Bunch'. Use a more complicated version that handles more edge cases correctly https://github.com/bashtage/linearmodels/blob/master/linearmodels/utility.py#L21 called `AttrDict` which describes its purpose.

File: statsmodels/tsa/vector_ar/var_model.py (hunk 21 above)
Comment (bashtage) [reply statsmodels/tsa/vector_ar/var_model.py]: I switches it to return.

**Comment by josef-pkt** [comment]
I did a quick skimming and spot checking.

Because VAR is old, the backwards compatibility restrictions and deprecation policy needs to be pretty strict. 
E.g. sigma_u still needs to be available with deprecation warning, e.g. as a property that delegates to cov_resid. I didn't see whether that is the case.
I didn't check `y` or other names that are being dropped.
(related aside: We do not support using positional arguments as keywords, so positional arguments can be renamed without problems.  (There are maybe a few exceptions where positional arguments are also allowed by keyword usage given popular usage, e.g. data in from_formula))

Also large refactorings should be kept in digestible pieces, and it's better to make one uncontroversial basic improvement, separate structural refactorings that have possibly a larger impact and chance for unintentionally breaking user code.
Based on my skimming so far, this PR looks pretty much in the uncontroversial basic refactoring style.

And, larger refactorings of established code is better merged at the beginning of a release cycle, so we have some time to fix refactoring problems, which show up almost always.

**Comment by bashtage** [comment]
> Also large refactorings should be kept in digestible pieces, and it's better to make one uncontroversial basic improvement, separate structural refactorings that have possibly a larger impact and chance for unintentionally breaking user code.

This is the digestible version -- it was one monolithic changeset until I sliced it into ~15 sets of grouped changes.

**Comment by jbrockmendel** [comment]
> Also large refactorings should be kept in digestible pieces

I definitely like the many-small-PRs approach, but the review/merge process around here is not conducive to it.

**Comment by jbrockmendel** [comment]
> These should all get done before the next release I would hope,

Agreed.  Is there a goal for when that will be?

File: statsmodels/tsa/vector_ar/var_model.py [hunk 24]
```diff
@@ -1815,31 +2440,11 @@ def test_inst_causality(self, causing, signif=0.05):
...
-        verbose : bool
-            If True, print a table with the results.
 
         Returns
         -------
-        results : dict
-            A dict holding the test's results. The dict's keys are:
-
-            "statistic" : float
-              The calculated test statistic.
```
[the remark this answers is not in the document]
Comment (josef-pkt) [reply statsmodels/tsa/vector_ar/var_model.py]: but that requires an extra step for users which is annoying if we just want to know what the main attributes are
e.g. https://github.com/statsmodels/statsmodels/pull/4829/files#diff-36a55e208f6d1db357321ee87ef3ab77R1285

File: statsmodels/tsa/vector_ar/irf.py [hunk 25]
```diff
@@ -296,11 +296,11 @@ def errband_mc(self, orth=False, svar=False, repl=1000,
         model = self.model
         periods = self.periods
         if svar == True:
-            return model.sirf_errband_mc(orth=orth, repl=repl, T=periods,
+            return model.sirf_errband_mc(orth=orth, repl=repl, horizon=periods,
                                         signif=signif, seed=seed,
                                         burn=burn, cum=False)
```
Comment (jbrockmendel) [inline statsmodels/tsa/vector_ar/irf.py]: Indent

File: statsmodels/tsa/vector_ar/irf.py (hunk 3 above)
Comment (bashtage) [reply statsmodels/tsa/vector_ar/irf.py]: Saved for irf.py cleanup

File: statsmodels/tsa/vector_ar/irf.py (hunk 4 above)
Comment (bashtage) [reply statsmodels/tsa/vector_ar/irf.py]: Probably, but saving for irf clean

File: statsmodels/tsa/vector_ar/tests/test_var.py (hunk 9 above)
Comment (bashtage) [reply statsmodels/tsa/vector_ar/tests/test_var.py]: I added this decorator with some extensions as part of this PR.  Saved a lot of error prone copy-paste pain upgrading attributes to properties.

**Comment by bashtage** [comment]
@josef-pkt I have verified that there are no attributes dropped in this PR.

File: statsmodels/tsa/vector_ar/tests/test_var.py (hunk 9 above)
Comment (jbrockmendel) [reply statsmodels/tsa/vector_ar/tests/test_var.py]: I like the edits you made, will incorporate them downstream.  There's also a test ready to be copied.

**Comment by jbrockmendel** [comment]
vecm, test_coint, and parse_jmulti_vecm_output have whitespace in blank lines.  This might be an opportunity to get rid of that without too much teeth-pulling.

**Comment by josef-pkt** [comment]
maybe `p` should be renamed to something informative (`cov_decomp` or `var_decomp ?  
`irf` has `var_decomp` as chol(Omega) where omega is now cov_resid, and var_decomp is the same as `p` AFAICS

related issue and possible code examples that should still work
Generalized Impulse Response Function
https://groups.google.com/d/msg/pystatsmodels/BqMqOIghN78/U46bIhz50H8J

**Comment by josef-pkt** [comment]
maybe `horizon` -> ` steps`, 
steps is used in the forecast functions.
ArmaProcess impulse response function, `impulse_response`, uses `leads` 
statespace MLEResults uses `impulse_responses` as method name with `step` for leads/horizon.

I like `steps` the best now, mainly because of the forecast similarity

File: statsmodels/tsa/vector_ar/var_model.py [hunk 26]
```diff
@@ -1611,12 +1612,16 @@ def irf_resim(self, orth=False, repl=1000, T=10,
...
 
-        def fill_coll(sim):
-            ret = VAR(sim, exog=self.exog).fit(maxlags=k_ar, trend=self.trend)
-            ret = ret.orth_ma_rep(maxn=T) if orth else ret.ma_rep(maxn=T)
-            return ret.cumsum(axis=0) if cum else ret
+        def fill_coll(simulation):
+            ret = VAR(simulation, exog=self.exog).fit(maxlags=k_ar,
+                                                      trend=self.trend)
+            if orth:
+                rep = ret.orth_ma_rep(maxn=horizon)
```
Comment (josef-pkt) [inline statsmodels/tsa/vector_ar/var_model.py]: pre-existing bug: orth_ma_rep is missing `p` (var_ or cov_decomp) option for generalized impulse response function

(Note, I only did some basic checking of the simulated methods for impulse response function when finishing up VECM merge. Not enough unit tests)

File: statsmodels/tsa/vector_ar/var_model.py [hunk 27]
```diff
@@ -926,22 +925,33 @@ class VARProcess(object):
...
+
+    where
+
+    .. math::
+
+       u_t \sim MVN(0, \Sigma_u)
     """
-    def __init__(self, coefs, coefs_exog, cov_resid, names=None, _params_info=None):
+    def __init__(self, coefs, coefs_exog, cov_resid, names=None,
+                 k_trend=None, k_exog_user=None):
```
Comment (josef-pkt) [inline statsmodels/tsa/vector_ar/var_model.py]: I had introduced `_params_info` to move away from having to keep the info about all individual pieces of params around. For example some models switched to using a Parameter class to have params handling in a central object.
VECM has also deterministic seasonal dummies which are not (yet) in VAR.

During VECM merge, only the basic exog support was added to VAR. The full exog and deterministic support in VAR still needs work.  One possibility is to merge trend, seasonal dummies into exog, but in VECM that has the problem that inference and stochastic properties are not the same for deterministic and weakly exogenous variables.

File: statsmodels/tsa/vector_ar/var_model.py (hunk 26 above)
Comment (bashtage) [reply statsmodels/tsa/vector_ar/var_model.py]: p is optional now as it was supposed to be. 

https://github.com/statsmodels/statsmodels/pull/5220/files#diff-e2d6024b9d0ed6c0410bf2fffc1f1254R1169

File: statsmodels/tsa/vector_ar/var_model.py [hunk 28]
```diff
@@ -913,24 +914,30 @@ def select_order(self, maxlags=None, trend="c"):
 
         return LagOrderResults(ics, selected_orders)
 
+    @property
+    def exog_names(self):
```
Comment (josef-pkt) [inline statsmodels/tsa/vector_ar/var_model.py]: why did you turn this into a property?

File: statsmodels/tsa/vector_ar/var_model.py [hunk 29]
```diff
@@ -1083,7 +1083,7 @@ def plotsim(self, steps=None, offset=None, seed=None):
         """Deprecated.  Use plot_sim."""
         import warnings
         warnings.warn("plotsim is deprecated and will be removed in 0.11.0. "
-                      "Use plot_sim.", DeprecationWarning)
+                      "Use plot_sim.", FutureWarning)
```
Comment (josef-pkt) [inline statsmodels/tsa/vector_ar/var_model.py]: Isn't DeprecationWarning in these cases more appropriate?

Comment (bashtage) [reply statsmodels/tsa/vector_ar/var_model.py]: There is now guidance from python.org: Future for users, Deprecation for developers.  So Future is almost always the right warning.  Older posts seemed to indicate Future was for this is definitely going away and deprecation is for things that may go away but shouldn't be used because there is a better way now.

File: statsmodels/tsa/vector_ar/var_model.py (hunk 28 above)
Comment (bashtage) [reply statsmodels/tsa/vector_ar/var_model.py]: Better docstring and to match endog_names

File: statsmodels/tsa/vector_ar/var_model.py (hunk 27 above)
Comment (bashtage) [reply statsmodels/tsa/vector_ar/var_model.py]: The real problem is that the hierarchy between VARProcess and VAR is wrong.  I don't think this is the PR to fix that issue.  It is better to get it cleaned up a lot and bring in some of the existing PRs rather than to get bogged down trying to do a big refactor at the same time.

Comment (josef-pkt) [reply statsmodels/tsa/vector_ar/var_model.py]: But whatever refactoring is still coming, IMO, the `_params_info` is more flexible and more future proof/resistant, than switching back to using individual k_xxx in the signature.

e.g. where do we put deterministic seasonal of VECM?
My guess is that we need or can use an `exog_all` that combines, trend/constant, exog_user and deterministic, for those places when they can all be treated in the same way.
However, for forecasting we can build trend and seasonal components directly, but users need to provide exog_user.

Comment (bashtage) [reply statsmodels/tsa/vector_ar/var_model.py]: VARProcess should know about trends that can be well defined, e.g., 'c' or 'ct', but general trend arrays should not be passed to a VARProcess since these aren't consistent with defining a 'process'.  A process should, for example, be simulable from its definition. There probably needs to be an intermediate, hidden class, after VARProcess and before other user-visible classes that add more general support of VAR-X type specifications.

Comment (josef-pkt) [reply statsmodels/tsa/vector_ar/var_model.py]: I don't see anything special about 'c' or 'ct' trend. Anything that changes the unconditional mean in a period can be lumped together for simulation.
e.g. we can have a predict_mean method that creates the exog/deterministic part which can include seasonal components, known structural breaks in trend, future exog that are policy variables and set outside of the model.
I don't see why the "process" would care about what drives the "strictly exogenous" mean shift in each period.

Comment (bashtage) [reply statsmodels/tsa/vector_ar/var_model.py]: 'nc','c', and 'ct' are special in the sense that they, in combination with model parameters and a distribution of the shocks, provide a complete description of a stochastic process.  A (nobs, q) array of exog "data" without the structure of a complete process behind it, do not.   IOW passing 'c' is providing an exog "generator' that can generate draws from the exogenous process for any number of time periods or across many simulations. These definitions allow the question as to whether the process is_stable to be correctly answered since the process is well defined. 

> I don't see why the "process" would care about what drives the "strictly exogenous" mean shift in each period.

This isn't a complete description of a stochastic process without another "process" the describes the exogenous shifts.

Comment (jbrockmendel) [reply statsmodels/tsa/vector_ar/var_model.py]: > But whatever refactoring is still coming, IMO, the _params_info is more flexible and more future proof/resistant, than switching back to using individual k_xxx in the signature.

I agree with @bashtage on this one (recall I objected to this pattern back when it was first implemented)

`_params_info` is functionally equivalent to `**kwargs`, with the main difference that it is non-standard so introduces cognitive overhead for non-Josef readers (and makes future maintenance more Josef-constrained).  It also does no validation on passed keys, a point on which the explicit version in this PR is strictly superior.

If this is really future-proof/resistant, why is this pattern not used elsewhere?  I don't like the `**kwargs` pattern either, but at least it is something readers are used to.

> e.g. where do we put deterministic seasonal of VECM?

IIRC from when this was merged, the _only_ extant tests are for deterministic seasonal cases.  These could be captured in a "trend" variable like "s4" or "s12".

Splitting `VARProcess` into `VARProcess` and `VARXProcess(VARProcess)` is not a topic for this PR, but is one that I would definitely support (thematically in line with #5206).  It is much simpler to get the "A process should, for example, be simulable from its definition." behavior without the special-casing within `VARProcess.__init__`

Comment (josef-pkt) [reply statsmodels/tsa/vector_ar/var_model.py]: In my view moving away from `_params_info` is a regression.
The main disadvantage of kwargs is that they are all main keywords instead of using separate dicts or instances of some classes by functional separation, e.g. I'm using cov_kwds and optim_kwds in some places.

AFAIK, nc, c, ct are only of theoretical interest because it is more easy to get some asymptotics.
seasonal dummies or seasonal splines are theoretically similar and treated as stationary deterministics in VECM.
In practice we also want to allow structural breaks in trend or non-power polynomials like chebycheff.
And we might want to do conditional simulation that takes a specific path of the `exog` as given.
E.g. when we want to simulate temperature specific endog, then conditioning on the termperature path of the last century might be a useful simulation.

At the end of VECM merge, there was still the open question how to define some of the time series properties in the presence of trend or exog, e.g. long-run mean doesn't make sense with trend or trending exog, and I think those process properties including impulse response functions should be defined ignoring all trend, deterministic and exog variables, but those are not ignored for long-term prediction at a finite forecast horizon.

**Comment by jbrockmendel** [comment]
Possibly closes #815?

**Comment by bashtage** [comment]
> Possibly closes #815?

I don't think so, at least not yet.

File: statsmodels/tsa/vector_ar/vecm.py [hunk 30]
```diff
@@ -1939,7 +1939,7 @@ def resid(self):
         return self.y_all.T[self.k_ar:] - self.fittedvalues
 
     def test_normality(self, signif=0.05):
-        """
+        r"""
         Test assumption of normal-distributed errors using Jarque-Bera-style
         omnibus :math:`\\chi^2` test.
```
Comment (jbrockmendel) [inline statsmodels/tsa/vector_ar/vecm.py]: Should there still be double \\?

**Comment by jbrockmendel** [comment]
@bashtage why did you close this?  The only potential problem I see is that there are tests for deprecated_alias usages that don't seem to be in place.

**Comment by bashtage** [comment]
There seemed to be real will to end up fix the API which seemed to be necessary to getting VAR to be coherent enough that it could be called correct.  So I found other tings to do.
